File: golang/internal/domain/stock.go
package domain

import (
	"errors"
	"fmt"

	"github.com/shopspring/decimal"
)

// Stock は保有銘柄（銘柄と保有額）を表す。
type Stock struct {
	Symbol    StockSymbol
	AmountJpy decimal.Decimal
}

type PortfolioItem struct {
	Symbol StockSymbol
	Rate   decimal.Decimal
}

// Portfolio は最適ポートフォリオ（銘柄ごとの構成比率）を表す。
type Portfolio struct {
	Items []PortfolioItem
}

func NewPortfolio(items []PortfolioItem) (Portfolio, error) {
	if len(items) == 0 {
		return Portfolio{}, errors.New("portfolio must have at least one item")
	}
	sum := decimal.Zero
	for _, item := range items {
		sum = sum.Add(item.Rate)
	}
	if !sum.Equal(decimal.NewFromInt(1)) {
		return Portfolio{}, fmt.Errorf("portfolio rates must sum to 1, got %s", sum.String())
	}
	seen := make(map[StockSymbol]struct{})
	for _, item := range items {
		if _, ok := seen[item.Symbol]; ok {
			return Portfolio{}, errors.New("portfolio must not have duplicate symbols")
		}
		seen[item.Symbol] = struct{}{}
	}
	return Portfolio{Items: items}, nil
}

File: typescript/src/domain/stock.ts
import Decimal from "decimal.js";
import { StockSymbol } from "./stockSymbol";

// Stock は保有銘柄（銘柄と保有額）を表す。
export interface Stock {
  symbol: StockSymbol;
  amountJpy: Decimal;
}

export interface PortfolioItem {
  symbol: StockSymbol;
  rate: Decimal;
}

// Portfolio は最適ポートフォリオ（銘柄ごとの構成比率）を表す。
export class Portfolio {
  readonly items: ReadonlyArray<PortfolioItem>;

  constructor(items: ReadonlyArray<PortfolioItem>) {
    if (items.length === 0) {
      throw new Error("portfolio must have at least one item");
    }
    const sum = items.reduce((acc, i) => acc.plus(i.rate), new Decimal(0));
    if (!sum.equals(1)) {
      throw new Error(`portfolio rates must sum to 1, got ${sum.toString()}`);
    }
    const symbols = new Set(items.map((i) => i.symbol));
    if (symbols.size !== items.length) {
      throw new Error("portfolio must not have duplicate symbols");
    }
    this.items = items;
  }
}
